ee/modules/021-cni-cilium/hooks/ee/types.go

```diff
@@ -0,0 +1,70 @@
+/*
+Copyright 2024 Flant JSC
+Licensed under the Deckhouse Platform Enterprise Edition (EE) license. See https://github.com/deckhouse/deckhouse/blob/main/ee/LICENSE
+*/
+
+package ee
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"sort"
+
+	"k8s.io/apimachinery/pkg/types"
+)
+
+var hashFunc func(key string) [32]byte
+
+func init() {
+	hashFunc = newHashFunc()
+}
+
+type EgressGatewayState struct {
+	Generation                  int64             `json:"generation"`
+	Name                        string            `json:"name"`
+	Mode                        string            `json:"mode"`
+	IP                          string            `json:"ip"`
+	RoutingTableName            string            `json:"routingTableName"`
+	UID                         types.UID         `json:"uid,omitempty"`
+	DesiredActiveNode           string            `json:"desiredActiveNode"`
+	AllNodes                    []string          `json:"allNodes"`
+	ReadyNodes                  []string          `json:"readyNodes"`
+	HealthyNodes                []string          `json:"healthyNodes"`
+	HealthyNodesWithEgressAgent []string          `json:"healthyNodesWithEgressAgent"`
+	CurrentActiveNodes          []string          `json:"currentActiveNodes"`
+	NodeSelector                map[string]string `json:"nodeSelector"`
+}
+
+func (eg *EgressGatewayState) electDesiredActiveNode() string {
+	if len(eg.ReadyNodes) == 0 {
+		return ""
+	}
+
+	// Algorithm: If the readyNodes slice is empty, then there is no sense to calculate desiredMaster.
+	// For each of readyNode, calculate the SHA256 hash of (EgressGatewayGroup.name + Node.Name).
+	// Add the hashes to an array and sort it. Select the first element from the resulting slice.
+
+	readyNodesCopy := make([]string, len(eg.ReadyNodes))
+	copy(readyNodesCopy, eg.ReadyNodes)
+
+	sort.Slice(readyNodesCopy, func(i, j int) bool {
+		hi := hashFunc(eg.Name + "#" + readyNodesCopy[i])
+		hj := hashFunc(eg.Name + "#" + readyNodesCopy[j])
+		return bytes.Compare(hi[:], hj[:]) < 0
+	})
+
+	eg.DesiredActiveNode = readyNodesCopy[0]
+	return readyNodesCopy[0]
+}
+
+func newHashFunc() func(key string) [32]byte {
+	hashedValues := make(map[string][32]byte)
```

Comment:
It seems like `hashedValues` is a cache without eviction policy. In clusters with dynamic nodes,  it will lead to a memory leak.